src/Interpreters/ActionsDAG.h

```diff
@@ -89,6 +96,10 @@ class ActionsDAG
         /// For COLUMN node and propagated constants.
         ColumnPtr column;
 
+        /// Determine if this action should be executed lazily. If it should and the action type is FUNCTION, then the function
+        /// won't be executed and will be stored with it's arguments in ColumnFunction with isShortCircuitArgument() = true.
+        LazyExecution lazy_execution = LazyExecution::DISABLED;
+
```

Comment:
IMHO, this is not used in ActionsDAG. So, should be moved to `ExpressionActions` (probably, `ExpressionActions::Action`)

Reply:
We should set `lazy_execution` for nodes before compiling expressions, that means before creating actions in linearizeActions. Moreover, we should check `lazy_execution` while compiling expressions to not compile short circuit function with lazy executed arguments. I had no choice to place it here.

Reply:
We can store a separate set of lazily executed nodes. And use it inside `ActionsDAG::compileFunctions`.
Also, maybe, `compileFunctions` should be moved to `ExpressionActions`.
If you think it is not reasonable, maybe we can merge this way, and refactor later.